src/modules/job-manager/update.c

```diff
@@ -0,0 +1,232 @@
+ * are.
+ *
+ * Once all updates are validated by callbacks, then updates as applied
+ * to jobspec are validated by passing an updated jobspec to the
+ * `job.validate` jobtap plugin stack.
+ *
+ * Plugins may request that this validation step be skipped by setting
+ * the 'validated' flag to 1 in the FLUX_PLUGIN_OUT_ARGS of the
+ * `job.update.*` callback. The `job.validate` step will only be skipped
+ * all keys in an update have the validated flag set.
+ *
+ * If all steps above are successful, then a `jobspec-update` upddate
+ * event is posted for the job and a success response sent to the
+ * caller.
+ *
+ * FUTURE WORK
+ *
+ * - Some job updates may require feasibility checks on the resulting
+ *   jobspec. There should be a flag for a plugin to require that the
+ *   result be passed to the scheduler feasibility RPC.
+ *
+ * - The above change will require some asynchronous handling be added
+ *   to this service
+ *
+ * - Plugins should also somehow be able to initiate asynchronous work
+ *   before validating an update. There is no support for async plugin
+ *   callbacks in jobtap at this time, though.
+ */
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+#include <flux/core.h>
+
+#include "update.h"
+#include "job-manager.h"
+#include "jobtap-internal.h"
+#include "event.h"
+
+struct update {
+    struct job_manager *ctx;
+    flux_msg_handler_t **handlers;
+};
+
+
+static int update_job (struct job_manager *ctx,
+                       struct flux_msg_cred cred,
+                       struct job *job,
+                       json_t *updates,
+                       char **errp)
+{
+    const char *key;
+    json_t *value;
+    int validate = 0; /* validation of result necessary */
+
+    /*  Loop through one or more proposed updates in `updates` object
+     *  and call `job.update.<key>` job plugin(s) to validate each
+     *  update.
+     */
+    json_object_foreach (updates, key, value) {
+        int needs_validation = 1;
+        if (jobtap_job_update (ctx->jobtap,
+                               cred,
+                               job,
+                               key,
+                               value,
+                               &needs_validation,
+                               errp) < 0)
+            return -1;
+        /*  If any jobspec key needs further validation, then all
+         *  keys will be validated at the same time. This means a key
+         *  that might not need further validation when updated alone
+         *  may need to be validated when paired with other keys in a
+         *  single update:
+         */
+        if (needs_validation)
+            validate = 1;
+    }
+    if (validate
+        && jobtap_validate_updates (ctx->jobtap,
+                                    job,
+                                    updates,
+                                    errp) < 0)
+            return -1;
+    /*  All updates have been allowed by pluginss and validated as a unit,
```

Comment:
s/pluginss/plugins/

Reply:
sheesh did the typos ci action not run?

Reply:
i thought the same thing, but i think these are all typos the typo checker is just missing

tried `~/bin/typos/typos src/modules/job-manager` and got nothing.

Reply:
should add to typos when we find them :-)  

https://github.com/crate-ci/typos/issues/784

Reply:
Ah, well I did a good job with my typos then. :-)